### PR TITLE
Display requested version when CPM enabled in dotnet list package

### LINKED ISSUE #12765: dotnet list package doesn't list requested versions when using CPM
### NuGet Product Used

dotnet.exe

### Product Version

8.0.0-preview.5.23280.8

### Worked before?

_No response_

### Impact

None

### Repro Steps & Context

On a project that uses CPM, `dotnet list package` (both format text and json have the same problem), the requested version is not listed:

```text
Project 'NuGet.PackageManagement.UI.Test' has the following package references
   [net472]:
   Top-level Package                                     Requested   Resolved
   > FluentAssertions                                                6.6.0
   > Microsoft.NET.Test.Sdk                                          17.1.0
   > Microsoft.VisualStudio.Sdk.TestFramework.Xunit                  17.2.7
   > Microsoft.VisualStudioEng.MicroBuild.Core                       1.0.0
   > Moq                                                             4.18.1
   > xunit                                                           2.4.2
   > xunit.runner.visualstudio                                       2.4.5
```

### Verbose Logs

_No response_

### PR DESCRIPTION
**Pull request by martinrrm**
## Bug

Fixes: https://github.com/NuGet/Home/issues/12765

Regression? Last working version:

## Description

`dotnet list package` retrieves the information from the project or asset files, since CPM disables the ability of using the `Version` property in `PackageReference` tag, `Requested` version information is missing when doing the command.

This PR checks if the version is CentrallyManaged and populates the `Requested` information from the Directory.Packages.props file instead of the project file.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: You're using `Any` here and then you're using `Where`. It's unnecessary.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: All packages are expected to be centrally managed, unless there's `VersionOverride`. Not sure if that check adds something.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Does `VersionOverride` work today?

**Review by nkolev92 — COMMENTED** [review]
How does VersionOverride work?

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (martinrrm) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Yess it works for CPM, in the Samples repo we have an example I have been using to test this

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (martinrrm) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: You're right, I can just check if the CPM is enabled. 

I did some testing and even if the user has `VersionOverride` it will appear as centrally managed since it's going to be declared in the Directory.Packages.props file too.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Does `dotnet list package` work with it? Does it display the correct version?

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (aortiz-msft) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: >tfmInformation.CentralPackageVersions.Where(cpv => cpv.Key.Equals(topLevelPackage.Name, StringComparison.Ordinal)).First(); [](http://example.com/codeflow?start=64&length=123)

Wouldn't this first iterate through the entire collection to find all matches, instantiating a new collection for them, then return the first element of the new collection? A for loop where we break at the first match would be more efficient.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs
Comment (aortiz-msft) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs:86]: Should we also add a test for a global reference?

**Review by aortiz-msft — CHANGES_REQUESTED** [review]
:clock1:

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (martinrrm) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: This is an example, the version in Directory.Packages.props is 13.0.0 and the VersionOverride is 9.0.1
image

Comment (nkolev92) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: That's incorrect in that case. We need to have requested be `9.0.1`.

Comment (martinrrm) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: I see your point and I believe its valid. I was looking at it the other way. In this example: For me I know that I'm requesting 13.0.0 but I want that my `VersionOverride` version is the one that gets Resolved. 

@jeffkl Can you help us clarify the expected behavior here?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: VersionOverride changes what you're requesting. 
It's basically overriding what the PackageVersion wanted.

Comment (martinrrm) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Thanks for the clarification! I'll change the behavior of that

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs
Comment (martinrrm) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs:86]: Talked with Jeff about GlobalReferences and the feature its implemented, I'll add tests for that

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Maybe we can add a visual indicator that there's a version override?  
Similar to say "Auto Referenced packages". 

I'm just brainstorming here, you can wait for Jeff to confirm :)

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs
Comment (aortiz-msft) [reply test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs:86]: I'm signing off now assuming these tests will be added before merging.

**Review by aortiz-msft — DISMISSED** [review]

**Review by aortiz-msft — DISMISSED** [review]
Thank you for adding the test for GlobalPackageReference.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (dtivel) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Does CPM require package ID's to be case sensitive?

Suppose in Directory.packages.props you have one casing (e.g.:  NuGet.Common), but in your project-level override you have a different casing (e.g.:  nuget.common).  Would CPM support this?  It looks like this line of code does not.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (dtivel) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: The value of `VersionOverride` can be `null`, which you handled.  However, it looks like [`MinVersion`](https://github.com/NuGet/NuGet.Client/blob/216cdfabbeefef8ce52d16d424a1a3cbc8a262d0/src/NuGet.Core/NuGet.Versioning/VersionRangeBase.cs#L27) can also be `null`, which is why [`HasLowerBound`](https://github.com/NuGet/NuGet.Client/blob/216cdfabbeefef8ce52d16d424a1a3cbc8a262d0/src/NuGet.Core/NuGet.Versioning/VersionRangeBase.cs#L42) exists.

Is it not possible to specify a [version range](https://github.com/NuGet/NuGet.Client/blob/216cdfabbeefef8ce52d16d424a1a3cbc8a262d0/src/NuGet.Core/NuGet.Versioning/VersionRangeBase.cs#L27) for `VersionOverride` or is it assumed that no one will?

```XML
<PackageReference Include="abc" VersionOverride="[,2.0)" />
```

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: All package ids are case insensitive, so this is for sure a bug.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: The rest of the code handles that by displaying the range. I'm guessing original version. We should follow that pattern.

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <TargetFramework>net8.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="NuGet.Common" Version="[,6.6.1]" />
    <PackageReference Include="NuGet.Commands" Version="6.5.0" />
    <PackageReference Include="NuGet.Protocol" Version="[6.5.0,)" />
  </ItemGroup>
</Project>

```

```console
PS C:\Code\Temp\DLP> dotnet list package 
Project 'DLP' has the following package references
   [net8.0]: 
   Top-level Package      Requested   Resolved
   > NuGet.Commands       6.5.0       6.5.0
   > NuGet.Common         [,6.6.1]    3.2.0
   > NuGet.Protocol       [6.5.0,)    6.5.0

```

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Package ID is case insensitive, so this comparer needs updated.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs
Comment (nkolev92) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs]: Readability would be improved if it's obvious the id/version that this method creates.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs
Comment (nkolev92) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs]: I feel like this is incorrect. 

VersionOverride overrides the requested version and the command should reflect.

**Review by nkolev92 — CHANGES_REQUESTED** [review]

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (martinrrm) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: You're right, I was under the impression that Ranges weren't allowed, floating are not allowed. 

I'll have to investigate this more, I was using `MinVersion` because `VersionRange.OriginalString` is always a range even for if you're not using one.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (dtivel) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Does [this existing statement](https://github.com/NuGet/NuGet.Client/pull/5350/files#diff-90c9045bb743a57a0e781f15088f964285ff4adf24e006ed4cd3a68d7e520674L754) have a similar bug?  With @martinrrm's changes, the line moved down a bit, but it's still intact.  I didn't scrutinize the code closely, but since we're talking about case insensitivity correctness, the surrounding code is up for discussion.  Not sure why it was case-sensitive to begin with.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: Yeah, I think that's a bug too.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (aortiz-msft) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: IIRC, CPM does not allow ranges overall because the intent is to give repo maintainers full control over which dependencies are used..

**Review by jeffkl — DISMISSED** [review]

**Comment by martinrrm** [comment]
I'm still trying to fix the tests 😢 I'm not 100% sure why its failing.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (aortiz-msft) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:654]: Missing new param documentation.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (aortiz-msft) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:659]: Where is IncludeProjects used?

**Comment by aortiz-msft** [comment]
>                 var projPackages = GetPackageReferencesFromTargets(projectPath, tfmInformation.ToString());

I know this is pre-existing, but we shouldn't abbreviate in variable names. Would you please rename to projectPackages?

---
Refers to: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:731 in 345ea37. [](commit_id = 345ea37ea6ca7730391f550d3e887e9d0164cb7b, deletion_comment = False)

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (aortiz-msft) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:754]: >var projectPackage = projPackages.Where(p => p.Name.Equals(topLevelPackage.Name, StringComparison.Ordinal)).First(); [](http://example.com/codeflow?start=32&length=116)

Why not use a foreach and break at the first result instead?

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (aortiz-msft) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: >its [](http://example.com/codeflow?start=67&length=3)

typo: replace "its" with "it's"

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (aortiz-msft) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs]: >topLevelPackage.VersionOverride?.OriginalStrin [](http://example.com/codeflow?start=67&length=46)

Would doing this check at the beginning save us work from lines 757 - 760?

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (aortiz-msft) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:879]: >VersionOverride [](http://example.com/codeflow?start=55&length=15)

I'm a bit concerned that these properties are not defined centrally somewhere. Would you please create a new issue for this?

Comment (martinrrm) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:879]: I created the issue https://github.com/NuGet/Home/issues/12865, assigned to myself for next quality week.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:756]: I feel like this is contradicting line 759 https://github.com/NuGet/NuGet.Client/pull/5350/files#diff-90c9045bb743a57a0e781f15088f964285ff4adf24e006ed4cd3a68d7e520674R759 which is handling verison overide cases.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:768]: This gets the version from versionoverride right?

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs
Comment (nkolev92) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs:87]: This test and the previous are technically the same so you might be able to use a Theory to avoid duplication. 

FWIW, I'm not sure how relevant it is to have both, having this one is probably more than sufficient.

File: test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs
Comment (nkolev92) [inline test/NuGet.Core.FuncTests/Dotnet.Integration.Test/DotnetListPackageTests.cs]:
```suggestion
public async Task DotnetListPackage_WithCPM_WithVersionOverride()
```

Correctly naming the test will help find tests with for VersionOverride easier.

**Review by nkolev92 — COMMENTED** [review]
Looks great. 

1 question and some suggestions for the tests.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:768]: FYI, I use github's website to look at PR comments, but I believe that you use VSCode. This is what I see: 
image

I assume you're referring to the code UNDER the line the comment is on, but github shows the 4 lines of code above. Your comment doesn't make sense with the info that github provides, so I'd like to ask you to please adapt to how github presents PR comments, so that it makes more sense for reviewers.

**Review by jeffkl — DISMISSED** [review]

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:768]: I was referring to the code below yeah. 

I'll try to remember that for the future.

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (martinrrm) [reply src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:756]: Removing lines 759 and its corresponding if block, VersionOverride is already handled like you pointed out below

File: src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs
Comment (jeffkl) [inline src/NuGet.Core/NuGet.CommandLine.XPlat/Utility/MSBuildAPIUtility.cs:651]:
```suggestion
        /// <param name="project">Project to get the resolved versions from</param>
```

**Review by jeffkl — APPROVED** [review]

**Review by nkolev92 — APPROVED** [review]

**Pull request merged**
